array.c

```diff
@@ -3488,60 +3551,52 @@ rb_ary_bsearch(VALUE ary)
 
 /*
  *  call-seq:
- *    array.bsearch_index {|element| ... } -> integer or nil
- *    array.bsearch_index -> new_enumerator
+ *    array.bsearch_index(target: :first) {|element| ... } -> integer or nil
+ *    array.bsearch_index(target: :first) -> new_enumerator
  *
  *  Searches +self+ as described at method #bsearch,
  *  but returns the _index_ of the found element instead of the element itself.
  */
 
 static VALUE
-rb_ary_bsearch_index(VALUE ary)
+rb_ary_bsearch_index(int argc, VALUE *argv, VALUE ary)
 {
-    long low = 0, high = RARRAY_LEN(ary), mid;
-    int smaller = 0, satisfied = 0;
-    VALUE v, val;
+    VALUE opt;
 
-    RETURN_ENUMERATOR(ary, 0, 0);
-    while (low < high) {
-        mid = low + ((high - low) / 2);
-        val = rb_ary_entry(ary, mid);
-        v = rb_yield(val);
-        if (FIXNUM_P(v)) {
-            if (v == INT2FIX(0)) return INT2FIX(mid);
-            smaller = (SIGNED_VALUE)v < 0; /* Fixnum preserves its sign-bit */
-        }
-        else if (v == Qtrue) {
-            satisfied = 1;
-            smaller = 1;
-        }
-        else if (!RTEST(v)) {
-            smaller = 0;
-        }
-        else if (rb_obj_is_kind_of(v, rb_cNumeric)) {
-            const VALUE zero = INT2FIX(0);
-            switch (rb_cmpint(rb_funcallv(v, id_cmp, 1, &zero), v, zero)) {
-              case 0: return INT2FIX(mid);
-              case 1: smaller = 0; break;
-              case -1: smaller = 1;
-            }
-        }
-        else {
-            rb_raise(rb_eTypeError, "wrong argument type %"PRIsVALUE
-                     " (must be numeric, true, false or nil)",
-                     rb_obj_class(v));
-        }
-        if (smaller) {
-            high = mid;
-        }
-        else {
-            low = mid + 1;
-        }
+    RETURN_ENUMERATOR_KW(ary, argc, argv, RB_PASS_KEYWORDS);
+
+    rb_scan_args(argc, argv, "0:", &opt);
+
+    if (NIL_P(opt)) {
+        return ary_bsearch_index(ary, true);
     }
-    if (!satisfied) return Qnil;
-    return INT2FIX(low);
-}
 
+    static ID keyword_ids[1];
+    VALUE target;
+    ID target_id;
+
+    if (!keyword_ids[0]) {
+        keyword_ids[0] = rb_intern_const("target");
+    }
+
+    rb_get_kwargs(opt, keyword_ids, 0, 1, &target);
+
+    if (!SYMBOL_P(target)) {
+        goto invalid_target;
+    }
+
+    target_id = SYM2ID(target);
+
+    if (target_id == rb_intern("first")) {
+        return ary_bsearch_index(ary, true);
+    }
+    else if (target_id == rb_intern("last")) {
+        return ary_bsearch_index(ary, false);
+    }
+
+  invalid_target:
+    rb_raise(rb_eArgError, "target must be :first or :last");
```

Comment:
Do not inadvertently convert an input to ID.
It can cause ID table overflow.

```suggestion
    if (target == ID2SYM(rb_intern("first"))) {
        return ary_bsearch_index(ary, true);
    }
    else if (target == ID2SYM(rb_intern("last"))) {
        return ary_bsearch_index(ary, false);
    }
    else {
        rb_raise(rb_eArgError, "target must be :first or :last");
        UNREACHABLE_RETURN(Qnil);
    }
```

Reply:
Thank you for your suggestion. I have fixed it.